src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -74,36 +91,59 @@ void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecora
   // client should use "" instead.
   assert(msg != nullptr, "enqueuing a null message!");
 
-  if (!_buffer->push_back(output, decorations, msg)) {
+  size_t msg_len = strlen(msg);
+
+  ConsumerLocker clocker;
+  if (_buffer->push_back(output, decorations, msg, msg_len)) {
+    _data_available = true;
+    clocker.notify();
+    return;
+  }
+
+  if (LogConfiguration::async_mode() == LogConfiguration::AsyncMode::Stall) {
+    size_t size = Message::calc_size(msg_len);
+    void* ptr = os::malloc(size, mtLogging);
```

Comment:
Your mention of recursion made me check if we are logging from our `os:malloc/free` calls. I can find at least one path  do `log_warning` in debug builds. This will cause both recursive locking on our PlatformMonitor and cause infinite recursion if the second message does not fit in the buffer. 

Maybe this logging is unreachable from this specific context. But I find it a little scary nonetheless. Maybe this should be documented somewhere.

Reply:
Hi Axel,

I've already switched to having the consumer thread log anything it itself logs directly without going through the buffer so this won't be an issue. See `enqueue_if_locked` (I'll be removing that and only have `enqueue` soon, however, so check there if you can't find it).

Reply:
Not sure I understand what you are trying to say. Maybe it will be clearer when you have pushed code.

But I am not talking about the consumer. I am talking about any thread which logs and calls into here, then fails to write to the buffer, and calls `os::malloc` which in turn may log, which calls back into this code. But maybe you are saying there are already guardrails around logging under the ProducerLock.